mysqlnd_azure.c

```diff
@@ -233,412 +233,459 @@ get_redirect_info(const MYSQLND_CONN_DATA * const conn, char* redirect_host, cha
+
+            //init redirect_conn options failed, just use the proxy connection
+            if (ret == FAIL) {
+                DBG_ENTER("[redirect]: init redirection option failed. ");
+                redirect_conn->m->dtor(redirect_conn); //release created resource
+
+                if(MYSQLND_AZURE_G(enableRedirect) == REDIRECT_ON) {
+                    //REDIRECT_ON, abort the original connection
+                    conn->m->send_close(conn);
+                    SET_CLIENT_ERROR(conn->error_info, MYSQLND_AZURE_ENFORCE_REDIRECT_ERROR_NO, UNKNOWN_SQLSTATE, "Connection aborted because init redirection failed.");
+                    goto err;
+                } else {
+                    //REDIRECT_PREFERRED, do nothing else for redirection, just use the previous connection
+                    goto after_conn;
+                }
+            }
 
-			}
-		}
+            //init redirect_conn succeeded, use this conn to start a new connection and handshake
+ 
+            const MYSQLND_CSTRING redirect_hostname = { redirect_host, strlen(redirect_host) };
+            const MYSQLND_CSTRING redirect_username = { redirect_user, strlen(redirect_user) };
+            MYSQLND_STRING redirect_transport = redirect_conn->m->get_scheme(redirect_conn, redirect_hostname, &socket_or_pipe, ui_redirect_port, &unix_socket, &named_pipe);
+
+            const MYSQLND_CSTRING redirect_scheme = { redirect_transport.s, redirect_transport.l };
+
+            enum_func_status redirectState = redirect_conn->m->connect_handshake(redirect_conn, &redirect_scheme, &redirect_username, &password, &database, mysql_flags);
+
+            if (redirectState == PASS) { //handshake with redirect_conn succeeded, replace original connection info with redirect_conn and add the redirect info into cache table
+                DBG_ENTER("[redirect]: mysql redirect handshake succeeded.");
+
+                //add the redirect info into cache table
+                mysqlnd_azure_add_redirect_cache(username.s, hostname.s, port, redirect_username.s, redirect_hostname.s, ui_redirect_port);
+
+                //close previous proxy connection
+                conn->m->send_close(conn);
+                conn->m->dtor(conn);
+                if (transport.s) {
+                    mnd_sprintf_free(transport.s);
+                    transport.s = NULL;
+                }
+
+                //upate conn, pfc,  pconn for later user
+                conn = redirect_conn;
+                pfc = redirect_conn->protocol_frame_codec; //this variable will be used in after_conn context, so need take care
+                *pconn = redirect_conn; //use new conn outside for caller
+
+                //upate host, user, transport for later user
+                hostname = redirect_hostname;
+                username = redirect_username;
+                port = ui_redirect_port;
+                transport = redirect_transport;
+                redirect_transport = NULL;
```

Comment:
Sorry, My previous comment might not clear, I was thinking we should set redirect_transport.s and redirect_transport.l to empty value, here we assign NULL to a structure,  I am not sure the behavior for different compilers, but we should avoid this. I suggest we just set the whole struct to 0.

Reply:
Thanks. You are right. After check, I think the safe way to set the whole struct to 0 need use memset, or copy from an empty struct, is this right?  I chose to only set redirect_transport.s=NULL instead.